lib/Transforms/StagedPatternRewriteDriver.cpp

```diff
@@ -0,0 +1,289 @@
+  /// know.
+  void notifyOperationRemoved(Operation *op) override {
+    LLVM_DEBUG({
+      logger.startLine() << "** OperationRemoved  : '" << op->getName() << "'("
+                         << op << ")\n";
+    });
+  }
+
+  // When a root is going to be replaced, its removal will be notified as well.
+  // So there is nothing to do here.
+  void notifyRootReplaced(Operation *op) override {
+    LLVM_DEBUG({
+      logger.startLine() << "** OperationRemoved  : '" << op->getName() << "'("
+                         << op << ")\n";
+    });
+    llvm_unreachable("Unsupported notifyRootReplaced");
+  }
+
+  /// This method is used to notify the rewriter that an in-place operation
+  /// modification is about to happen. A call to this function *must* be
+  /// followed by a call to either `finalizeRootUpdate` or `cancelRootUpdate`.
+  /// This is a minor efficiency win (it avoids creating a new operation and
+  /// removing the old one) but also often allows simpler code in the client.
+  virtual void startRootUpdate(Operation *op) override {
+    LLVM_DEBUG({
+      logger.startLine() << "** StartRootUpdate  : '" << op->getName() << "'("
+                         << op << ")\n";
+    });
+  }
+
+  /// This method is used to signal the end of a root update on the given
+  /// operation. This can only be called on operations that were provided to a
+  /// call to `startRootUpdate`.
+  virtual void finalizeRootUpdate(Operation *op) override {
+    LLVM_DEBUG({
+      logger.startLine() << "** FinalizeRootUpdate  : '" << op->getName()
+                         << "'(" << op << ")\n";
+    });
+    addToWorklist(op, worklist, worklistMap);
+  }
+
+public:
+#ifndef NDEBUG
+  /// A logger used to emit information during the application process.
+  llvm::ScopedPrinter logger{llvm::dbgs()};
+#endif
+
+private:
+  /// Keep track of the round number.
+  int64_t roundNum = 0;
+  /// The low-level pattern applicator.
+  PatternApplicator matcher;
+  /// Keep track of patterns applied by this rewriter to only apply them
+  /// the specified number of times (i.e. once for now).
+  llvm::SmallDenseSet<const Pattern *> appliedPatterns;
+
+  /// List of newly created ops at each round.
+  /// The worklist for this transformation keeps track of the operations that
+  /// need to be revisited, plus their index in the worklist.  This allows us to
+  /// efficiently remove operations from the worklist when they are erased, even
+  /// if they aren't the root of a pattern.
+  std::vector<Operation *> worklist;
+  DenseMap<Operation *, unsigned> worklistMap;
+};
+
+} // namespace
+
+LogicalResult
+mlir::applyStagedPatterns(ArrayRef<Operation *> roots,
+                          const FrozenRewritePatternSet &stage1Patterns,
+                          const FrozenRewritePatternSet &stage2Patterns,
+                          function_ref<LogicalResult(FuncOp)> stage3Lambda) {
+  if (roots.empty())
+    return failure();
```

Comment:
I'd say this is success. There is just nothing to do with no roots, which never produces errors.